identity_credential/src/validator/domain_linkage_validator.rs

```diff
@@ -0,0 +1,530 @@
+            source: None,
+          })?;
+        }
+      }
+      OneOrMany::One(_) => {
+        Err(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+      }
+    };
+
+    match &credential.credential_subject {
+      OneOrMany::One(credential_subject) => {
+        match &credential_subject.id {
+          None => {
+            return Err(DomainLinkageValidationError {
+              cause: DomainLinkageValidationErrorCause::MissingSubjectId,
+              source: None,
+            });
+          }
+          Some(id) => match CoreDID::parse(id.to_string()) {
+            Ok(subject_did) => {
+              if issuer_did != subject_did {
+                return Err(DomainLinkageValidationError {
+                  cause: DomainLinkageValidationErrorCause::IssuerSubjectMismatch,
+                  source: None,
+                });
+              }
+            }
+            Err(_) => {
+              return Err(DomainLinkageValidationError {
+                cause: DomainLinkageValidationErrorCause::InvalidSubjectId,
+                source: None,
+              });
+            }
+          },
+        }
+
+        let origin: &str = credential_subject
+          .properties
+          .get("origin")
+          .ok_or(DomainLinkageValidationError {
+            cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?
+          .as_str()
+          .ok_or(DomainLinkageValidationError {
+            cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
+            source: None,
+          })?;
+
+        let origin_url: Url = match Url::parse(origin) {
+          Ok(url) => url,
+          Err(identity_core::Error::InvalidUrl(url::ParseError::RelativeUrlWithoutBase)) => {
+            match Url::parse("https://".to_owned() + origin) {
+              Ok(url) => url,
+              Err(err) => {
+                return Err(DomainLinkageValidationError {
+                  cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
+                  source: Some(Box::new(err)),
+                });
+              }
+            }
+          }
+          Err(err) => {
+            return Err(DomainLinkageValidationError {
+              cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
+              source: Some(Box::new(err)),
+            });
+          }
+        };
+
+        if origin_url.origin() != domain.origin() {
+          return Err(DomainLinkageValidationError {
+            cause: DomainLinkageValidationErrorCause::OriginMismatch,
+            source: None,
+          });
+        }
+      }
+      OneOrMany::Many(_) => {
+        return Err(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::MultipleCredentialSubjects,
+          source: None,
+        });
+      }
+    }
```

Comment:
I find it easier to read this if it is split up a bit more with some comments explaining the intention of each block. Here is a suggestion:

```suggestion
 // Extract credential subject
    let OneOrMany::One(ref credential_subject) = credential.credential_subject else {
      return Err(DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::MultipleCredentialSubjects,
        source: None,
      });
    };

    // Validate credential subject
    {
      let subject_id = credential_subject.id.as_deref().ok_or(DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::MissingSubjectId,
        source: None,
      })?;

      let subject_did = CoreDID::parse(subject_id.as_str()).map_err(|_| DomainLinkageValidationError {
        cause: DomainLinkageValidationErrorCause::InvalidSubjectId,
        source: None,
      })?;

      if issuer_did != subject_did {
        return Err(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::IssuerSubjectMismatch,
          source: None,
        });
      }
    }

    // Extract and validate origin
    {
      let origin: &str = credential_subject
        .properties
        .get("origin")
        .and_then(|value| value.as_str())
        .ok_or(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
          source: None,
        })?;

      let origin_url: Url = match Url::parse(origin) {
        Ok(url) => Ok(url),
        Err(identity_core::Error::InvalidUrl(url::ParseError::RelativeUrlWithoutBase)) => {
          Url::parse("https://".to_owned() + origin).map_err(|err| DomainLinkageValidationError {
            cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
            source: Some(Box::new(err)),
          })
        }
        Err(other_error) => Err(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::InvalidSubjectOrigin,
          source: Some(Box::new(other_error)),
        }),
      }?;

      if origin_url.origin() != domain.origin() {
        return Err(DomainLinkageValidationError {
          cause: DomainLinkageValidationErrorCause::OriginMismatch,
          source: None,
        });
      }
    }
```

Reply:
I think that's great, thanks for the comments, too!